File: utils/gcs_utils.js
/**
 * Shared GCS (Google Cloud Storage) utility functions for all components
 */

const { Storage } = require('@google-cloud/storage');
const fs = require('fs');
const path = require('path');

/**
 * Upload a file to Google Cloud Storage
 * 
 * @param {string} localFilePath - Path to the local file to upload
 * @param {string} destinationPath - Destination path in GCS bucket (e.g., 'dvb/2026-01-28/article.txt')
 * @returns {Promise<Object>} Upload status information
 */
async function uploadFileToGCS(localFilePath, destinationPath) {
    const bucketName = process.env.GCS_BUCKET;
    
    if (!bucketName) {
        console.log('⚠️  GCS_BUCKET environment variable not set');
        console.log('⚠️  Skipping GCS upload');
        return { status: 'skipped', reason: 'GCS_BUCKET not set' };
    }

    try {
        // Initialize GCS client (uses service account from Cloud Run)
        const storage = new Storage();
        const bucket = storage.bucket(bucketName);
        
        // Upload file
        await bucket.upload(localFilePath, {
            destination: destinationPath,
            metadata: {
                contentType: 'text/plain',
            },
        });
        
        const fileSize = fs.statSync(localFilePath).size;
        const folderPath = path.posix.dirname(destinationPath);
        const fileUri = `gs://${bucketName}/${destinationPath}`;
        const folderUri = `gs://${bucketName}/${folderPath === "." ? "" : `${folderPath}/`}`;

        console.log(`✅ File uploaded to ${fileUri}`);
        console.log(`   Saved folder: ${folderUri}`);
        
        return {
            status: 'success',
            bucket: bucketName,
            folder: folderPath,
            filename: destinationPath,
            size_bytes: fileSize,
            url: fileUri
        };
        
    } catch (error) {
        console.error(`❌ Error uploading to GCS: ${error.message}`);
        console.error(`   Bucket: ${bucketName}`);
        throw error;
    }
}

/**
 * Upload JSON data directly to GCS
 * 
 * @param {Object} data - JavaScript object to save as JSON
 * @param {string} destinationPath - Destination path in GCS bucket
 * @returns {Promise<Object>} Upload status information
 */
async function uploadJSONToGCS(data, destinationPath) {
    const bucketName = process.env.GCS_BUCKET;
    
    if (!bucketName) {
        console.log('⚠️  GCS_BUCKET environment variable not set');
        console.log('⚠️  Skipping GCS upload');
        return { status: 'skipped', reason: 'GCS_BUCKET not set' };
    }

    try {
        // Initialize GCS client
        const storage = new Storage();
        const bucket = storage.bucket(bucketName);
        
        // Convert to JSON string
        const jsonPayload = JSON.stringify(data, null, 2);
        
        // Upload JSON
        const blob = bucket.file(destinationPath);
        await blob.save(jsonPayload, {
            contentType: 'application/json',
        });

        const folderPath = path.posix.dirname(destinationPath);
        const fileUri = `gs://${bucketName}/${destinationPath}`;
        const folderUri = `gs://${bucketName}/${folderPath === "." ? "" : `${folderPath}/`}`;

        console.log(`✅ JSON uploaded to ${fileUri}`);
        console.log(`   Saved folder: ${folderUri}`);
        
        return {
            status: 'success',
            bucket: bucketName,
            folder: folderPath,
            filename: destinationPath,
            size_bytes: Buffer.byteLength(jsonPayload, 'utf8'),
            url: fileUri
        };
        
    } catch (error) {
        console.error(`❌ Error uploading JSON to GCS: ${error.message}`);
        console.error(`   Bucket: ${bucketName}`);
        throw error;
    }
}

/**
 * Upload text content directly to GCS
 * 
 * @param {string} content - Text content to upload
 * @param {string} destinationPath - Destination path in GCS bucket
 * @returns {Promise<Object>} Upload status information
 */
async function uploadTextToGCS(content, destinationPath) {
    const bucketName = process.env.GCS_BUCKET;
    
    if (!bucketName) {
        console.log('⚠️  GCS_BUCKET environment variable not set');
        console.log('⚠️  Skipping GCS upload');
        return { status: 'skipped', reason: 'GCS_BUCKET not set' };
    }

    try {
        // Initialize GCS client
        const storage = new Storage();
        const bucket = storage.bucket(bucketName);
        
        // Upload text
        const blob = bucket.file(destinationPath);
        await blob.save(content, {
            contentType: 'text/plain',
        });

        const folderPath = path.posix.dirname(destinationPath);
        const fileUri = `gs://${bucketName}/${destinationPath}`;
        const folderUri = `gs://${bucketName}/${folderPath === "." ? "" : `${folderPath}/`}`;

        console.log(`✅ Text uploaded to ${fileUri}`);
        console.log(`   Saved folder: ${folderUri}`);
        
        return {
            status: 'success',
            bucket: bucketName,
            folder: folderPath,
            filename: destinationPath,
            size_bytes: Buffer.byteLength(content, 'utf8'),
            url: fileUri
        };
        
    } catch (error) {
        console.error(`❌ Error uploading text to GCS: ${error.message}`);
        console.error(`   Bucket: ${bucketName}`);
        throw error;
    }
}

module.exports = {
    uploadFileToGCS,
    uploadJSONToGCS,
    uploadTextToGCS
};

File: utils/gcs_utils.py
"""
Shared GCS (Google Cloud Storage) utility functions for all components
"""

import os
import json
from datetime import datetime
import torch
from google.cloud import storage


def save_results_to_gcs(results_data, confirm_save=True):
    """
    Save computation results to Google Cloud Storage.
    
    Args:
        results_data (dict): Results data to save
        confirm_save (bool): If True, actually save to GCS. If False, only simulate/validate.
    
    Returns:
        dict: Upload status information
    """
    print("\n" + "=" * 60)
    print("SAVING RESULTS TO GCS")
    print("=" * 60)
    
    bucket_name = os.environ.get("GCS_BUCKET")
    if not bucket_name:
        print("⚠️  GCS_BUCKET environment variable not set")
        print("⚠️  Skipping GCS upload")
        return {"status": "skipped", "reason": "GCS_BUCKET not set"}
    
    # Create unique filename with timestamp
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    job_name = os.environ.get("JOB_NAME", "unknown-job")
    filename = f"{job_name}/results_{timestamp}.json"
    
    # Add metadata
    results_data["timestamp"] = datetime.now().isoformat()
    results_data["job_name"] = job_name
    results_data["cuda_available"] = torch.cuda.is_available()
    
    if torch.cuda.is_available():
        results_data["gpu_name"] = torch.cuda.get_device_name(0)
        results_data["gpu_count"] = torch.cuda.device_count()
    
    # Prepare JSON payload
    json_payload = json.dumps(results_data, indent=2)
    
    if not confirm_save:
        print("🔍 DRY RUN MODE - Not actually saving to GCS")
        print(f"Would save to: gs://{bucket_name}/{filename}")
        print(f"Payload size: {len(json_payload)} bytes")
        print(f"Preview:\n{json_payload[:200]}...")
        return {
            "status": "dry_run",
            "bucket": bucket_name,
            "filename": filename,
            "size_bytes": len(json_payload)
        }
    
    try:
        # Initialize GCS client (uses service account from Cloud Run)
        client = storage.Client()
        bucket = client.bucket(bucket_name)
        
        # Upload to GCS
        blob = bucket.blob(filename)
        blob.upload_from_string(
            json_payload,
            content_type="application/json"
        )
        
        folder_name = os.path.dirname(filename)
        file_url = f"gs://{bucket_name}/{filename}"
        folder_url = f"gs://{bucket_name}/{folder_name}/" if folder_name else f"gs://{bucket_name}/"

        print(f"✅ Results saved to {file_url}")
        print(f"   Saved folder: {folder_url}")
        
        return {
            "status": "success",
            "bucket": bucket_name,
            "folder": folder_name,
            "filename": filename,
            "size_bytes": len(json_payload),
            "url": file_url
        }
        
    except Exception as e:
        print(f"❌ Error saving to GCS: {e}")
        print(f"   Bucket: {bucket_name}")
        raise
